// npm-registry-client/lib/initialize.js
var crypto = require('crypto')
var HttpAgent = require('http').Agent
var HttpsAgent = require('https').Agent

var pkg = require('../package.json')

var httpAgent
var httpsAgents = {}

module.exports = initialize

function initialize (uri, ssl, method, accept, headers) {
  if (!this.config.sessionToken) {
    this.config.sessionToken = crypto.randomBytes(8).toString('hex')
    this.log.verbose('request id', this.config.sessionToken)
  }

  var opts = {
    url: uri,
    method: method,
    headers: headers,
    localAddress: this.config.proxy.localAddress,
    strictSSL: ssl.strict,
    cert: ssl.certificate,
    key: ssl.key,
    ca: ssl.ca,
    agent: getAgent(uri.protocol, this.config, ssl)
  }

  // allow explicit disabling of proxy in environment via CLI
  //
  // how false gets here is the CLI's problem (it's gross)
  if (this.config.proxy.http === false) {
    opts.proxy = null
  } else {
    // request will not pay attention to the NOPROXY environment variable if a
    // config value named proxy is passed in, even if it's set to null.
    var proxy
    if (uri.protocol === 'https:') {
      proxy = this.config.proxy.https
    } else {
      proxy = this.config.proxy.http
    }
    if (typeof proxy === 'string') opts.proxy = proxy
  }

  headers.version = this.version || pkg.version
  headers.accept = accept

  if (this.refer) headers.referer = this.refer

  headers['npm-session'] = this.config.sessionToken
  headers['user-agent'] = this.config.userAgent

  return opts
}

function getAgent (protocol, config, ssl) {

  if (protocol === 'https:') {
    var key = JSON.stringify(ssl)

    var agent = httpsAgents[key];

    if (!agent) {
      agent = httpsAgents[key] = new HttpsAgent({
        keepAlive: true,
        localAddress: config.proxy.localAddress,
        rejectUnauthorized: ssl.strict,
        ca: ssl.ca,
        cert: ssl.certificate,
        key: ssl.key
      })
    }

    return agent
  } else {
    if (!httpAgent) {
      httpAgent = new HttpAgent({
        keepAlive: true,
        localAddress: config.proxy.localAddress
      })
    }

    return httpAgent
  }
}
